core/field_colour.js

```diff
@@ -107,44 +109,100 @@ Blockly.FieldColour.prototype.setValue = function(colour) {
   if (this.sourceBlock_) {
     // Set the primary, secondary and tertiary colour to this value.
     // The renderer expects to be able to use the secondary color as the fill for a shadow.
-    this.sourceBlock_.setColour(colour, colour, colour);
+    this.sourceBlock_.setColour(colour, colour, this.sourceBlock_.getColourTertiary());
   }
+  if (!opt_fromSliders) {
+    this.updateSliderHandles_();
+  }
+  this.updateDom_();
 };
 
 /**
- * Get the text from this field.  Used when the block is collapsed.
- * @return {string} Current text.
+ * Create the hue, saturation or value CSS gradient for the slide backgrounds.
+ * @param {string} channel – Either "hue", "saturation" or "value".
+ * @return {string} Array color hex color stops for the given channel
  */
-Blockly.FieldColour.prototype.getText = function() {
-  var colour = this.colour_;
-  // Try to use #rgb format if possible, rather than #rrggbb.
-  var m = colour.match(/^#(.)\1(.)\2(.)\3$/);
-  if (m) {
-    colour = '#' + m[1] + m[2] + m[3];
+Blockly.FieldColour.prototype.createColorStops_ = function(channel) {
+  var hsv = goog.color.hexToHsv(this.getValue());
+  var stops = [];
+  for(var n = 0; n <= 360; n += 20) {
+    switch (channel) {
+      case 'hue':
+        stops.push(goog.color.hsvToHex(n, hsv[1], hsv[2]));
+        break;
+      case 'saturation':
+        stops.push(goog.color.hsvToHex(hsv[0], n / 360, hsv[2]));
+        break;
+      case 'brightness':
+        stops.push(goog.color.hsvToHex(hsv[0], hsv[1], 255 * n / 360));
+        break;
+    }
   }
-  return colour;
+  return stops;
+};
+
+/**
+ * Set the gradient CSS properties for the given node and channel
+ * @param {Node} node - The DOM node the gradient will be set on.
+ * @param {string} channel – Either "hue", "saturation" or "value".
+ */
+Blockly.FieldColour.prototype.setGradient_ = function(node, channel) {
+  var stops = this.createColorStops_(channel);
+  goog.style.setStyle(node, 'background',
+        '-moz-linear-gradient(left, ' + stops.join(',') + ')');
+  goog.style.setStyle(node, 'background',
+        '-webkit-linear-gradient(left, ' + stops.join(',') + ')');
+  goog.style.setStyle(node, 'background',
+        '-o-linear-gradient(left, ' + stops.join(',') + ')');
+  goog.style.setStyle(node, 'background',
+        '-ms-linear-gradient(left, ' + stops.join(',') + ')');
+  goog.style.setStyle(node, 'background',
+        'linear-gradient(left, ' + stops.join(',') + ')');
```

Comment:
perhaps consider constructing this part left, ' + stops.join(',') + ') and the beginning of the function. 
Something like this? goog.style.setStyle(node, 'background', '-moz-linear-gradient(' + gradient;
I'm not actually sure how much you save by not doing the join over and over again though.